File: backend/domains/task/handlers/create_task.handler.go
package handlers

import (
	"context"

	"github.com/cloudwego/hertz/pkg/app"
	"github.com/erweixin/go-genai-stack/backend/domains/task/http/dto"
)

// CreateTaskHandler 创建任务（HTTP 适配层）
//
// 用例：CreateTask（参考 usecases.yaml）
//
// HTTP:
//   - Method: POST
//   - Path: /api/tasks
//
// Handler 职责（瘦层）：
//  1. 解析 HTTP 请求
//  2. 转换 DTO（使用转换层）
//  3. 调用 Domain Service
//  4. 转换响应（使用转换层）
//  5. 处理错误
//
// 业务逻辑在 service.TaskService.CreateTask() 中实现
func (deps *HandlerDependencies) CreateTaskHandler(ctx context.Context, c *app.RequestContext) {
	// 1. 获取用户 ID
	userID, exists := c.Get("user_id")
	if !exists {
		c.JSON(401, dto.ErrorResponse{
			Error:   "UNAUTHORIZED",
			Message: "未授权访问",
		})
		return
	}
	userIDStr, ok := userID.(string)
	if !ok {
		c.JSON(500, dto.ErrorResponse{
			Error:   "INTERNAL_ERROR",
			Message: "用户 ID 类型错误",
		})
		return
	}

	// 2. 解析 HTTP 请求
	var req dto.CreateTaskRequest
	if err := c.BindAndValidate(&req); err != nil {
		c.JSON(400, dto.ErrorResponse{
			Error:   "INVALID_INPUT",
			Message: "请求参数无效",
			Details: err.Error(),
		})
		return
	}

	// 3. 转换为 Domain Input（使用转换层）
	input, err := toCreateTaskInput(userIDStr, req)
	if err != nil {
		handleDomainError(c, err)
		return
	}

	// 4. 调用 Domain Service
	output, err := deps.taskService.CreateTask(ctx, input)
	if err != nil {
		handleDomainError(c, err)
		return
	}

	// 5. 转换为 HTTP 响应（使用转换层）
	c.JSON(200, toCreateTaskResponse(output))
}

File: backend-nodejs/domains/task/handlers/create_task.handler.ts
/**
 * CreateTask Handler
 * HTTP 适配层：处理创建任务的 HTTP 请求
 */

import type { FastifyRequest, FastifyReply } from 'fastify';
import type { HandlerDependencies } from './dependencies.js';
import type { CreateTaskRequest } from '../http/dto/task.js';
import { toCreateTaskInput, toCreateTaskResponse } from './converters.js';
import { requireUserId } from '../../../infrastructure/middleware/auth.js';
import { createContextFromRequest } from '../../../shared/types/context.js';

/**
 * CreateTask Handler
 * HTTP 适配层：处理创建任务的 HTTP 请求
 * 负责日志记录和性能监控，使用 Fastify 的 request.log（自动包含请求上下文）
 * 错误由全局错误处理中间件统一处理
 */
export async function createTaskHandler(
  deps: HandlerDependencies,
  req: FastifyRequest<{ Body: CreateTaskRequest }>,
  reply: FastifyReply
): Promise<void> {
  const startTime = Date.now();
  const userId = requireUserId(req);

  // 使用 Fastify 的 request.log（自动包含 traceId, requestId）
  req.log.info(
    {
      userId,
      title: req.body.title,
      priority: req.body.priority,
      hasDueDate: !!req.body.due_date,
      tagCount: req.body.tags?.length || 0,
    },
    'Creating task'
  );

  try {
    // 转换为 Domain Input
    const input = toCreateTaskInput(userId, req.body);

    // 创建请求上下文
    const ctx = createContextFromRequest({ userId, ...req });

    // 调用 Domain Service（纯业务逻辑，不包含日志）
    const output = await deps.taskService.createTask(ctx, input);

    // 记录成功日志和性能指标
    const duration = Date.now() - startTime;
    req.log.info(
      {
        taskId: output.task.id,
        userId: output.task.userId,
        duration,
      },
      'Task created successfully'
    );

    // 慢操作警告
    if (duration > 1000) {
      req.log.warn(
        {
          taskId: output.task.id,
          duration,
          threshold: 1000,
        },
        'Slow task creation'
      );
    }

    // 转换为 HTTP 响应
    reply.code(200).send(toCreateTaskResponse(output.task));
  } catch (error) {
    // 记录错误日志
    const duration = Date.now() - startTime;
    req.log.error(
      {
        userId,
        title: req.body.title,
        duration,
        error: error instanceof Error ? error.message : String(error),
        stack: error instanceof Error ? error.stack : undefined,
      },
      'Failed to create task'
    );
    throw error; // 由全局错误处理器处理
  }
}
